File: app/src/main/java/com/victoria/bleled/data/remote/NetworkObserver.java
package com.victoria.bleled.data.remote;

import android.content.Context;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import com.victoria.bleled.R;
import com.victoria.bleled.app.MyApplication;
import com.victoria.bleled.data.remote.myservice.BaseResponse;
import com.victoria.bleled.util.CommonUtil;
import com.victoria.bleled.util.arch.network.NetworkResult;

public class NetworkObserver<T> implements Observer<NetworkResult<T>> {
    private Context context = null;
    private boolean showError = true;

    public NetworkObserver() {
        this.showError = false;
        this.context = MyApplication.Companion.getGlobalApplicationContext();
    }

    public NetworkObserver(Context context, boolean showError) {
        this.context = context;
        this.showError = showError;
    }

    @Override
    public void onChanged(NetworkResult<T> result) {
        if (result == null) {
            return;
        }

        if (result.status.getValue() == NetworkResult.Status.success && result.data instanceof BaseResponse) {
            BaseResponse data = (BaseResponse) result.data;
            int error = data.getResult();
            if (error != ApiException.SUCCESS) {
                result.error = new ApiException(error, data.getMsg(), data.getReason());
                result.status = new MutableLiveData<>(NetworkResult.Status.error);
            }
        }

        // 재시도 등 UI처리가 필요할때 여기서 공통처리 진행
        if (result.status.getValue() == NetworkResult.Status.error && showError) {
            onError(result);
        }
    }

    protected void onError(NetworkResult<T> result) {
        if (result.error != null && result.error instanceof ApiException) {
            ApiException exception = (ApiException) result.error;
            CommonUtil.showToast(context, exception.getMsg() != null ? exception.getMsg() : context.getResources().getString(R.string.server_problem));
        } else {
            CommonUtil.showToast(context, R.string.network_connect_error);
        }
    }

    public String getErrStr(NetworkResult<T> result) {
        return getErrorMsg(context, result);
    }

    public static <T> Throwable getErrorException(NetworkResult<T> result) {
        if (result.status.getValue() == NetworkResult.Status.success && result.data instanceof BaseResponse) {
            BaseResponse data = (BaseResponse) result.data;
            int error = data.getResult();
            if (error != ApiException.SUCCESS) {
                result.error = new ApiException(error, data.getMsg(), data.getReason());
                result.status = new MutableLiveData<>(NetworkResult.Status.error);
            }
        }
        return result.error;
    }

    public static <T> String getErrorMsg(Context p_context, NetworkResult<T> result) {
        Throwable error = getErrorException(result);
        if (result.error != null && result.error instanceof ApiException) {
            ApiException exception = (ApiException) result.error;
            return exception.getMsg() != null ? exception.getMsg() : p_context.getResources().getString(R.string.server_problem);
        } else {
            return p_context.getResources().getString(R.string.network_connect_error);
        }
    }

}

File: app/src/main/java/com/victoria/bleled/data/remote/NetworkObserver.kt
package com.victoria.bleled.data.remote

import android.content.Context
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.victoria.bleled.R
import com.victoria.bleled.app.MyApplication.Companion.globalApplicationContext
import com.victoria.bleled.data.remote.myservice.response.BaseResp
import com.victoria.bleled.util.CommonUtil

open class NetworkObserver<T> : Observer<NetworkResult<T>> {
    private var context: Context? = null
    private var showError = true

    constructor() {
        showError = false
        context = globalApplicationContext
    }

    constructor(context: Context?, showError: Boolean) {
        this.context = context
        this.showError = showError
    }

    override fun onChanged(result: NetworkResult<T>) {
        if (result == null) {
            return
        }
        if (result.status.value == NetworkResult.Status.success && result.data is BaseResp) {
            val data = result.data as BaseResp
            val error = data.result
            if (error != ApiException.Companion.SUCCESS) {
                result.error = ApiException(error, data.message, data.reason)
                result.status = MutableLiveData(NetworkResult.Status.error)
            }
        }

        // 재시도 등 UI처리가 필요할때 여기서 공통처리 진행
        if (result.status.value == NetworkResult.Status.error && showError) {
            onError(result)
        }
    }

    private fun onError(result: NetworkResult<T>) {
        if (result.error != null && result.error is ApiException) {
            val exception = result.error as ApiException
            CommonUtil.showToast(
                context,
                if (exception.msg != null && !exception.msg!!.isEmpty()) exception.msg else context!!.resources.getString(
                    R.string.server_problem
                )
            )
        } else {
            CommonUtil.showToast(context, R.string.network_connect_error)
        }
    }

    fun getErrStr(result: NetworkResult<T>): String? {
        return getErrorMsg(context, result)
    }

    companion object {
        fun <T> getErrorException(result: NetworkResult<T>): Throwable {
            if (result.status.value == NetworkResult.Status.success && result.data is BaseResp) {
                val data = result.data as BaseResp
                val error = data.result
                if (error != ApiException.Companion.SUCCESS) {
                    result.error = ApiException(error, data.message, data.reason)
                    result.status = MutableLiveData(NetworkResult.Status.error)
                }
            }
            return result.error
        }

        fun <T> getErrorMsg(p_context: Context?, result: NetworkResult<T>): String? {
            val error = getErrorException(result)
            return if (result.error != null && result.error is ApiException) {
                val exception = result.error as ApiException
                if (exception.msg != null) exception.msg else p_context!!.resources.getString(R.string.server_problem)
            } else {
                p_context!!.resources.getString(R.string.network_connect_error)
            }
        }
    }
}